### ISSUE TITLE
Susceptibility distortion correction introduces distortion 

### ISSUE THREAD

**Issue by ChristianNSchmitz** [nipreps/fmriprep#3013]
### What happened?

Dear community,
we are using FMRIPREP for the preprocessing of task-based fMRI data. Apparently, the brain mask is not very accurate and the fMRI data is even more distorted after the susceptibility distortion correction (see image attached).
Thank you very much for your help!

### What command did you use?

```shell
docker run -ti --rm  \
  -u $(id -u):$(id -g) \
  -v /data/dataPreproc/rawdata:/data:ro \
  -v /data/dataPreproc/derivatives/fmriprep-preprocessing/:/out \
  -v /data/dataPreproc/workingDirectory/:/workDir \
  -v /home/christian.schmitz/license.txt:/opt/freesurfer/license.txt \
  nipreps/fmriprep:latest /data /out  \
  participant --participant-label "${sub}" \
    --output-spaces MNI152NLin6Asym:res-2 MNI152NLin2009cAsym fsLR \
    -w /workDir \
    --return-all-components \
    --cifti-output \
    --mem 52000;
```

### What version of fMRIPrep are you running?

23.0.2

### How are you running fMRIPrep?

Docker

### Is your data BIDS valid?

Yes

### Are you reusing any previously computed results?

No

### Additional information / screenshots

![sub-ZI004_ses-02_task-reversal_run-01_desc-sdc_bold](https://github.com/nipreps/fmriprep/assets/35763466/edc634c2-83d9-47c8-aa84-b33f785ff56b)
![Screenshot 2023-05-23 011624](https://github.com/nipreps/fmriprep/assets/35763466/3417063f-c0f2-4359-8025-24c456b082fc)

**Comment by ChristianNSchmitz** [comment]
I didn't flag SyN-SDC and used the fieldmap-based correction (default)

**Comment by effigies** [comment]
I see. The `--force-syn` made me think you were using SyN-SDC. What type of fieldmaps do you have, what are the orientations and PhaseEncodingDirections of all of your fieldmaps and BOLD files?

**Comment by ChristianNSchmitz** [comment]
Dear Chris, thanks for your help! We use a double-echo gradient echo field map sequence. 
Here are the basic properties of the sequences and I have attached the *.json files:
orientations:

Fieldmap:
PhaseEncodingDirections: "i",
"ImageOrientationPatientDICOM": [0.999333, -0.0201548, 0.0304505, 0.0235506, 0.993014, -0.115626],
 "ImageOrientationText": "Tra>Cor(-6.7)>Sag(1.6)"

Functional:
PhaseEncodingDirections: "j-",
"ImageOrientationPatientDICOM": [0.999333, -0.0201548, 0.0304505, 0.0235505, 0.993014, -0.115626],
"ImageOrientationText": "Tra>Cor(-6.7)>Sag(1.6)",



[sub-ZI004_ses-01_run-01_phasediff.txt](https://github.com/nipreps/fmriprep/files/11556568/sub-ZI004_ses-01_run-01_phasediff.txt)
[sub-ZI004_ses-01_task-examplerun-01_bold.txt](https://github.com/nipreps/fmriprep/files/11556569/sub-ZI004_ses-01_task-examplerun-01_bold.txt)

**Comment by ChristianNSchmitz** [comment]
Brief supplement: Also shutting off the slice time correction resulted in the same distored preprocessing

**Comment by marcelzwiers** [comment]
We have seen the same, however, with the latest version (23.1.#) it works well if you skull-strip / brain-mask the phasediff images before running fmriprep

**Comment by amirhusseinab** [comment]
Just as an additional update:
I have the same issue with both 23.0.2 and 23.1.3 versions on resting-state data applying _fieldmapless_ method in a data without fieldmap and preprocessing with the routine pipeline with the fieldmap data available. Of note, I am running fmriprep using Singularity on a HPC.

**Comment by JohannesWiesner** [comment]
I have the same issue (probably using the same settings as @ChristianNSchmitz, I did not acquire the dataset though) with `23.1.0`:

![Recording 2023-07-05 at 16 56 39](https://github.com/nipreps/fmriprep/assets/48254290/dbfe7dda-d79d-4594-99e8-3124b85f1221)

**Comment by themeo** [comment]
I have a very similar problem with the (almost) newest version (23.1.2; and older ones as well). I'm using the approach with fieldmaps estimated using two SBRef images with opposing orientations (AP PA). Running `topup`/`applytopup` manually on the same source images produces a correct solution. 

Original image (raw functional run)
![image](https://github.com/nipreps/fmriprep/assets/2484158/65fb93c8-0fec-4995-ae36-92623822cb0c)

topup-corrected:
![image](https://github.com/nipreps/fmriprep/assets/2484158/5b17bc16-67d8-4f58-84a9-82abbf9ae588)

fmriprep-corrected:
![image](https://github.com/nipreps/fmriprep/assets/2484158/85e58b55-a64b-4a02-b862-a107849c2005)




fmriprep was run using the following command:
`apptainer run --cleanenv --bind \$TMPDIR:/tmp /opt/fmriprep/23.1.2/fmriprep-23.1.2.simg /project/3011213.01/prederr/bids /project/3011213.01/prederr/bids/derivatives/fmriprep participant -w \$TMPDIR/sub-s32 --participant-label s32  --fs-license-file /opt_host/fmriprep/license.txt --mem_mb 20000 --omp-nthreads 4 --nthreads 4  --output-spaces anat func MNI152NLin2009cAsym --dummy-scans 5  --fs-subjects-dir /project/3011213.01/prederr/bids/derivatives/fastsurfer`

All images (both fieldmaps and BOLD file) had LAS orientation.
AP fieldmap and BOLD had "PhaseEncodingDirection": "j-"
PA fieldmap had "PhaseEncodingDirection": "j"

Like in @ChristianNSchmitz 's case, the BOLD was collected using a CMRR multiband sequence.

**Comment by effigies** [comment]
From discussion with @oesteban, the issue here is not the distortion correction but the Jacobian attenuation. These stretches outside the brain also happen with TOPUP (in the most recent post you can clearly see the shape of these stretches in both TOPUP and fMRIPrep), but are scaled by the Jacobian of the B-spline field.

Tracking this feature in https://github.com/nipreps/sdcflows/issues/382.

**Comment by tiborst** [comment]
I have encountered a similar problem in my dataset and I was wondering whether it might be possible to introduce already distortion corrected images into the fmriprep pipeline and use the tag --ignore fieldmap? 
Running fieldmap correction / unwarping in FSL did not introduce similar artefacts.

Would this be a feasible approach?

**Comment by amirhusseinab** [comment]
@tiborst I am now going for the same solution. There is another ongoing discussion [here](https://neurostars.org/t/using-synthetic-field-map-from-synbold-disco-with-fmriprep/25191), which you may be also interested in.

**Comment by andreifoldes** [comment]
> We have seen the same, however, with the latest version (23.1.#) it works well if you skull-strip / brain-extract the phasediff images before running fmriprep (use the magnitude image to compute the mask, e.g. with synthstrip)

@marcelzwiers could you please further details how you do the skullstripping, I'd like to give this a try.

**Comment by andreifoldes** [comment]
could it be possible make optionally available 20.2.7 LTS SDC pipeline in future releases of fmriprep?

**Comment by marcelzwiers** [comment]
> > We have seen the same, however, with the latest version (23.1.#) it works well if you skull-strip / brain-extract the phasediff images before running fmriprep (use the magnitude image to compute the mask, e.g. with synthstrip)
> 
> @marcelzwiers could you please further details how you do the skullstripping, I'd like to give this a try.

Sure, the basic idea is to compute the mask on the fieldmap magnitude image and apply that mask to the phasediff images in the fieldmap folder (overwriting the original phasediff data, so you may want to make a backup). In bidscoin this that be done with a single command, e.g.:
```console
skullstrip myproject/bids fmap/*_magnitude1* -m fmap/*_phasediff -o extra_data fmap
```
This save a masked `magnitude1` image in the `extra_data` subfolder and a masked `phasediff` image in the `fmap` folder
See here for more info:
https://bidscoin.readthedocs.io/en/latest/bidsapps.html#skull-stripping

**Comment by effigies** [comment]
@themeo Would you mind sharing the exact `topup`/`applytopup` commands you're using? It would be helpful to have a ground truth for comparing our fixes.

**Comment by themeo** [comment]
Sure! Nothing fancy here, though: 

`fslmerge -t AP_PA_b0.nii.gz sub-s21_acq-ROPESBRef_dir-AP_run-1_epi.nii.gz sub-s21_acq-ROPESBRef_dir-PA_run-1_epi.nii.gz`

`topup --imain=AP_PA_b0.nii.gz --datain=acqparams.txt --config=b02b0.cnf --out=topup_res --iout=AP_PA_b0_unwarped.nii.gz --fout=myfieldmap.nii.gz`

`applytopup --imain=sub-s21_task-Main_dir-AP_run-1_bold.nii.gz --inindex=1 --method=jac --datain=acqparams.txt --topup=topup_res --out=sub-s21_task-Main_dir-AP_run-1_desc-unwarped_bold.nii.gz`

I'd be happy to share specific data files if needed.

**Comment by effigies** [comment]
Thank you very much. That was helpful for debugging. I have a working fix in nipreps/sdcflows#391.

**Comment by amirhusseinab** [comment]
Thanks Chris! It solved my long-standing issue with SDC, but I encountered some problems running it. I tried to manually edit transform.py based on your solution with singularity --sandbox and rebuilding the singularity image. I did it on v23.1.4 and "fmriprep:next" versions.

- on v23.1.4. the analysis stopped, with an error in the node "Node Name: fmriprep_23_1_wf.single_subject_{subid}_wf.func_preproc_ses_001_task_rest_dir_RL_wf.unwarp_wf.resample" with "	AttributeError: 'B0FieldTransform' object has no attribute 'xfm'"
- on the fmriprep:next it ran smoothly, with the SDC applied (Thanks!), but no files in other spaces (T1w, fsLR, MNI*) are generated in the func folder, with no errors.

I know it is still in the development phase but I wondered if I could try to have things run since I'm facing a deadline ahead.

**Comment by effigies** [comment]
Let me see if I can backport these fixes without too much difficulty. I'm not sure how much the current fixes depend on larger changes we've made.

**Comment by effigies** [comment]
Please see #397 and #398. Tests from users experiencing the problems with blip-up/blip-down (#397) and phasediff/fieldmaps (#398) would be greatly appreciated. Instructions for testing with `fmriprep-docker` are in each PRs initial post.

**Comment by jameswyngaarden** [comment]
Hi all, thank you for all of the helpful information in this thread. Following up here from #3093--it looks like the fix in https://github.com/nipreps/sdcflows/pull/391 addresses the issue of abnormal distortions around the outlines of the image. However, we are also seeing strange distortions inside the corrected images for multiband=6, multi-echo=4 images, even when distortions around the outlines have been avoided: 

 Corrected:
<img width="966" alt="Screenshot 2023-10-30 at 9 00 24 AM" src="https://github.com/nipreps/fmriprep/assets/68560550/9f0712aa-cf1b-4ec6-a40f-74632a57c0e2">

Distorted:
<img width="961" alt="Screenshot 2023-10-30 at 9 00 51 AM" src="https://github.com/nipreps/fmriprep/assets/68560550/f55f4317-203f-45c1-8619-eaa86be55f40">


In this case, using fieldmap-free distortion correction in fmriprep 23.1.4 similarly resolves the issues around the outlines of the image, but we still see this abnormal pattern inside. I wanted to follow up to see if others have thoughts about addressing this pattern? Thank you!

```
	echo singularity run --cleanenv \
	-B ${TEMPLATEFLOW_DIR}:/opt/templateflow \
	-B ${MPLCONFIGDIR_DIR}:/opt/mplconfigdir \
	-B $maindir:/base \
	-B ~/work/tools/licenses:/opts \
	-B $scratchdir:/scratch \
	~/work/tools/fmriprep-23.1.4.simg \
	/base/bids /base/derivatives/fmriprep-syn \
	participant --participant_label $sub \
	--stop-on-first-crash \
	--me-output-echos \
	--ignore fieldmaps \
	--use-syn-sdc \
	--cifti-output 91k \
	--output-spaces fsLR fsaverage MNI152NLin6Asym \
	--bids-filter-file /base/code/fmriprep_config.json \
	--fs-license-file /opts/fs_license.txt -w /scratch >> $logdir/cmd_fmriprep_${PBS_JOBID}.txt
```

**Comment by manzouri** [comment]
Hi @effigies , I wonder if the fix for this will be implemented in next release before end of the year?

**Comment by effigies** [comment]
This fix is available in [23.2.0a1](https://github.com/nipreps/fmriprep/discussions/3139). Please test and let us know how it goes!

**Comment by jameswyngaarden** [comment]
Hi all, wanted to follow up now that we have tested the latest pre-release. In the previous version of fmriprep, fieldmap-free distortion correction looked pretty good for most sequences, but we have noticed now that use of the synthetic fieldmap in the latest version looks a bit odd. Here are screenshots from the same subject & sequence I posted about in #3093. I just wanted to flag this and note that we’re seeing this difference:
![image](https://github.com/nipreps/fmriprep/assets/68560550/85df81bc-1d83-43b0-a523-83522960090d)
![image (1)](https://github.com/nipreps/fmriprep/assets/68560550/612abd81-c747-479e-b7ed-0ff11072452e)
